torch/_functorch/aot_autograd.py

```diff
@@ -256,309 +539,145 @@ def from_fun(t):
-        for i, maybe_aliased_updated_inp in enumerate(maybe_inputs_with_mutated_metadata):
-            if maybe_aliased_updated_inp is None:
-                collect_metadata_mutation_input_info.append(None)
-                continue
-            # Figure out where the sizes/strides/storage_offset are in the compiled fw output.
-            sizes_idx = start_idx_for_aliased_output_metadata
-            strides_idx = sizes_idx + len(maybe_aliased_updated_inp.size())
-            storage_offset_idx = strides_idx + len(maybe_aliased_updated_inp.stride())
-            # update our offset for the next tensor
-            start_idx_for_aliased_output_metadata = storage_offset_idx + 1
-            inp_info = InputAliasInfo(
-                base_idx=i,
-                sizes_idx=sizes_idx,
-                strides_idx=strides_idx,
-                storage_offset_idx=storage_offset_idx,
-                tensor_meta=maybe_aliased_updated_inp,
-            )
-            collect_metadata_mutation_input_info.append(inp_info)
-
-        # Next, gather the metadata info on the user's outputs that alias (either inputs or graph outputs)
-        num_non_input_aliased_outputs = 0
-        for o in outs:
-            maybe_alias_info = aliased_out_to_inp_idx.get(o, None) if isinstance(o, torch.Tensor) else None
-            if maybe_alias_info is None:
-                output_type = OutputType.non_alias
-                # Here, alias_idx will tell us which output from the inner forward this corresponds to.
-                alias_idx = num_non_input_aliased_outputs
-                sizes_idx = None
-                strides_idx = None
-                storage_offset_idx = None
-                tensor_meta = None
-            else:
-                input_alias_idx, is_alias_of_intermediate_not_input, is_exact_input = maybe_alias_info
-                if is_exact_input:
-                    assert not is_alias_of_intermediate_not_input
-                    output_type = OutputType.alias_of_input
-                    alias_idx = input_alias_idx
-                    sizes_idx = None
-                    strides_idx = None
-                    storage_offset_idx = None
-                    tensor_meta = None
+            # TO FILL IN
+            elif isinstance(o, torch.Tensor) and o._base is not None and o.requires_grad and o._base.requires_grad:
+                # First, check if o's ._base is an existing output
+                output_type = OutputType.alias_of_intermediate
+                maybe_existing_out_idx = out_tensor_ids.get(id(o._base), None)
+                if maybe_existing_out_idx is not None:
+                    # the compiled forward returns (mutated_inps, outs),
+                    # so to index into the base correctly, we need to offset by # mutated inps
+                    base_idx = num_mutated_inps + maybe_existing_out_idx
```

Comment:
If you take my NamedTuple organization suggestion for the calling convention, then we can also restructure indexing operations into a tuple `(FieldEnum, idx)`. So for example, you want to represent an index to the offset here, so you'd say `(Out, maybe_existing_out_idx)`. No more manual offsetting; the Out says go to the out field, and then you do a normal index.

Reply:
Actually, all of this info can just be encoded in the existing `OutputType` enum. I'll end up with more invalid states if I introduce a separate FieldEnum. So the full enum will look something like:
```
OutputType = Enum(
    "OutputType", (
        "non_alias",
        "alias_of_input",
        "is_input",
        "alias_of_intermediate",
        "alias_of_intermediate_base_is_user_output",
        "alias_of_intermediate_save_as_output",
    )
)
```

Where:

"alias_of_intermediate": instructs the runtime code around `CompiledFunction.apply` to regenerate the output off of `graph_intermediates[base_idx]

"alias_of_intermediate_base_is_user_output": instructs the runtime code around `CompiledFunction.apply` to regenerate the output off of `user_outputs[base_idx]. (This is an edge case, where an output is a view of a graph intermediate, but that intermediate itself is already a user output).

"alias_of_intermediate_save_as_output": instructs the runtime code around `CompiledFunction.apply` to regenerate the output off of `graph_intermediates[base_idx]. Also instructs the code tracing the joint forward/backward to add `output._base` as a forward graph output.

Reply:
I think that in general there's a bit of tension here between precomputing stuff, and making the code easier to read. Technically, the existing code right now is more optimal, because we precompute the exact offset into `graph_outputs` to get the base, although this makes it harder to understand.

I'm instead going to update it so that that `base_idx` corresponds. to an index into either `user_outs` or `graph_intermediates`, and we instead construct an offset into the giant `graph_outputs` list at runtime, depending on `OutputType`.